File: src/infinicore/ops/sqrt/sqrt.cc
#include "infinicore/ops/sqrt.hpp"

namespace infinicore::op {

common::OpDispatcher<Sqrt::schema> &Sqrt::dispatcher() {
    static common::OpDispatcher<Sqrt::schema> dispatcher_;
    return dispatcher_;
}

void Sqrt::execute(Tensor output, Tensor input) {
    dispatcher().lookup(context::getDevice().getType())(output, input);
}

Tensor sqrt(Tensor input) {
    auto output = Tensor::empty(input->shape(), input->dtype(), input->device());
    sqrt_(output, input);
    return output;
}

void sqrt_(Tensor output, Tensor input) {
    Sqrt::execute(output, input);
}

} // namespace infinicore::op

File: python/infinicore/ops/sqrt.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor
import infinicore


def sqrt(input, *, out=None):
    if infinicore.use_ntops and input.device.type in ("cuda", "musa"):
        return infinicore.ntops.torch.sqrt(input, out=out)
    
    if out is None:
        return Tensor(_infinicore.sqrt(input._underlying))

    _infinicore.sqrt_(out._underlying, input._underlying)
    return out